File: litfass-server/src/main/scala/lit/fass/server/http/HttpServer.scala
package lit.fass.server.http

import akka.actor.typed.ActorSystem
import akka.actor.typed.scaladsl.adapter._
import akka.http.scaladsl.Http
import akka.http.scaladsl.server.Route

import scala.util.{Failure, Success}

/**
 * @author Michael Mair
 */
@Deprecated
object HttpServer {
  def startHttpServer(routes: Route, system: ActorSystem[_]): Unit = {
    // Akka HTTP still needs a classic ActorSystem to start
    implicit val classicSystem: akka.actor.ActorSystem = system.toClassic
    import system.executionContext

    val futureBinding = Http().bindAndHandle(routes, "localhost", 8080)
    futureBinding.onComplete {
      case Success(binding) =>
        val address = binding.localAddress
        system.log.info("Server online at http://{}:{}/", address.getHostString, address.getPort)
      case Failure(ex) =>
        system.log.error("Failed to bind HTTP endpoint, terminating system", ex)
        system.terminate()
    }
  }
}

File: litfass-server/src/main/kotlin/lit/fass/server/http/HttpServer.kt
package lit.fass.server.http

import akka.actor.typed.ActorSystem
import akka.http.javadsl.Http
import akka.http.javadsl.model.StatusCodes
import akka.http.javadsl.server.AllDirectives
import akka.http.javadsl.server.ExceptionHandler
import akka.http.javadsl.server.Route
import lit.fass.server.config.yaml.ConfigException
import lit.fass.server.execution.ExecutionException
import lit.fass.server.flow.FlowException
import lit.fass.server.logger
import lit.fass.server.persistence.PersistenceException
import lit.fass.server.retention.RetentionException
import lit.fass.server.schedule.SchedulerException

/**
 * @author Michael Mair
 */
class HttpServer(private val route: Route) : AllDirectives() {

    companion object {
        private val log = this.logger()
    }

    fun startHttpServer(system: ActorSystem<*>) {

        val exceptionHandler = ExceptionHandler.newBuilder()
            .match(ConfigException::class.java) { complete(StatusCodes.BAD_REQUEST, it.message) }
            .match(ExecutionException::class.java) { complete(StatusCodes.BAD_REQUEST, it.message) }
            .match(FlowException::class.java) { complete(StatusCodes.BAD_REQUEST, it.message) }
            .match(PersistenceException::class.java) { complete(StatusCodes.BAD_REQUEST, it.message) }
            .match(RetentionException::class.java) { complete(StatusCodes.BAD_REQUEST, it.message) }
            .match(SchedulerException::class.java) { complete(StatusCodes.BAD_REQUEST, it.message) }
            .match(Exception::class.java) { complete(StatusCodes.INTERNAL_SERVER_ERROR, it.message) }.build()

        val http = Http.get(system)
        http.newServerAt("0.0.0.0", 8080)
            .bind(handleExceptions(exceptionHandler) {
                route
            })
            .handle { serverBinding, throwable ->
                if (throwable == null) {
                    val address = serverBinding.localAddress()
                    log.info("Server online at http://${address.hostString}:${address.port}")
                } else {
                    log.error("Failed to bind HTTP endpoint, terminating system", throwable)
                    system.terminate()
                }
            }
    }
}
